### PR TITLE
Feature/pace 31 does a loose match on tag values

### PR DESCRIPTION
**Pull request by bvdeenen**
This is so that the same policy can be used on platforms with different tag rules (upper case, lower case, dashes etc.).
With loose matching `_`, ` ` and `-` are considered equal, as well as that it's a case insensitive match.

Also changed:
* `RefAndType` is gone
* `GlobalTransform.ref` is gone. Instead the transform type specific content is being used (currently only `TagTransform.tag_content`). `ref` in the database points to the type dependent primary key.

### CODE REVIEW

File: app/src/main/kotlin/com/getstrm/pace/config/AppConfiguration.kt [hunk 1]
```diff
@@ -7,6 +7,7 @@ import org.springframework.boot.context.properties.ConfigurationProperties
 data class AppConfiguration(
     val defaultViewSuffix: String = "",
     val catalogs: List<CatalogConfiguration> = emptyList(),
+    val looseTagMatch: Boolean = true,
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/config/AppConfiguration.kt]: The configuration might need to be more hierarchical to be able to add specific other configuration properties in the future. What about `globalTransforms.tagTransforms.looseTagMatch`?

File: app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt [hunk 2]
```diff
@@ -63,3 +78,8 @@ class GlobalTransformsDao(
             ).execute()
     }
 }
+
+/**
+ * returns SQL like style loose match. Underscore is any character.
+ */
+private fun RefAndType.toLooseMatch() = this.ref.replace(" ", "_").replace("-", "_")
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt]: Note that this requires the tags to be modified upon fetch like this as well. Otherwise you'll be comparing `a-tag` (stored in the DB) with `a_tag` (while `a-tag` was retrieved from the processing platform field), which does not match.

File: app/src/test/kotlin/com/getstrm/pace/dao/GlobalTransformsDaoTest.kt [hunk 3]
```diff
@@ -92,27 +95,83 @@ class GlobalTransformsDaoTest : AbstractDatabaseTest() {
...
+                .setRef(s)
+                .setTagTransform(TagTransform.newBuilder()
+                        .setTagContent(s)
+                        .addTransforms(FieldTransform.Transform.newBuilder().setNullify(getDefaultInstance()))
+                )
+                .build()
+
+        underTest.upsertTransform(transform).toGlobalTransform() shouldBe transform
+        underTest.listTransforms().size shouldBe 3
+        run {
```
Comment (trietsch) [inline app/src/test/kotlin/com/getstrm/pace/dao/GlobalTransformsDaoTest.kt]: Unnecessary `run` block.

File: app/src/test/kotlin/com/getstrm/pace/dao/GlobalTransformsDaoTest.kt [hunk 4]
```diff
@@ -92,27 +95,83 @@ class GlobalTransformsDaoTest : AbstractDatabaseTest() {
...
+                .setRef(s)
+                .setTagTransform(TagTransform.newBuilder()
+                        .setTagContent(s)
+                        .addTransforms(FieldTransform.Transform.newBuilder().setNullify(getDefaultInstance()))
+                )
+                .build()
+
+        strictDao.upsertTransform(transform).toGlobalTransform() shouldBe transform
+        strictDao.listTransforms().size shouldBe 3
+        run {
```
Comment (trietsch) [inline app/src/test/kotlin/com/getstrm/pace/dao/GlobalTransformsDaoTest.kt]: Same.

**Review by trietsch — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/getstrm/pace/config/AppConfiguration.kt (hunk 1 above)
Comment (ivan-p92) [reply app/src/main/kotlin/com/getstrm/pace/config/AppConfiguration.kt]: Sounds good, then you can also inject a more specific configuration class at the usage site than `AppConfiguration`

File: app/src/test/kotlin/com/getstrm/pace/dao/GlobalTransformsDaoTest.kt (hunk 3 above)
Comment (ivan-p92) [reply app/src/test/kotlin/com/getstrm/pace/dao/GlobalTransformsDaoTest.kt]: I suggest we stick to the Given When Then structure that we already use.

File: app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt (hunk 2 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt]: I don't agree.
What is stored is what is in the Request message, what is retrieved is what was stored. It's only a loose retrieval match. Global transforms are found via the tag value (from the platform) with uppercase, lowercase, dashes or whatever. If one is found (via the loose match), its transforms are added to the ruleset. The tag value itself is then no longer used.

File: app/src/main/kotlin/com/getstrm/pace/config/AppConfiguration.kt (hunk 1 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/config/AppConfiguration.kt]: done

File: app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt (hunk 2 above)
Comment (trietsch) [reply app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt]: Let's have a call.

**Review by trietsch — APPROVED** [review]

File: protos/getstrm/pace/api/global_transforms/v1alpha/global_transforms.proto [hunk 5]
```diff
@@ -68,7 +67,11 @@ message UpsertGlobalTransformResponse {
 }
 
 message DeleteGlobalTransformRequest {
```
Comment (trietsch) [inline protos/getstrm/pace/api/global_transforms/v1alpha/global_transforms.proto:74]: As the delete request now only uses primitives, you can modify the REST API to:
```
    option (google.api.http) = {
      delete: "/global-transforms/{ref}/{type}"
    };
```

File: protos/getstrm/pace/api/global_transforms/v1alpha/global_transforms.proto [hunk 6]
```diff
@@ -68,7 +67,11 @@ message UpsertGlobalTransformResponse {
...
-  repeated entities.v1alpha.GlobalTransform.RefAndType ref_and_types = 1;
+  string ref = 1;
```
Comment (trietsch) [inline protos/getstrm/pace/api/global_transforms/v1alpha/global_transforms.proto:75]: The `ref` does need a comment, that it's content refers to the primary keys of each `oneof`.

File: protos/getstrm/pace/api/global_transforms/v1alpha/global_transforms.proto [hunk 7]
```diff
@@ -35,7 +35,6 @@ service GlobalTransformsService {
   }
 }
 
-// As we also want to support a REST API and gRPC->JSON cannot map to messages, we cannot reuse RefAndType here.
 message GetGlobalTransformRequest {
   string ref = 1;
```
Comment (trietsch) [inline protos/getstrm/pace/api/global_transforms/v1alpha/global_transforms.proto:44]: The `ref` does need a comment, that it's content refers to the primary keys of each `oneof`.

File: protos/getstrm/pace/api/entities/v1alpha/entities.proto [hunk 8]
```diff
@@ -179,29 +179,28 @@ message DataPolicy {
...
-  string ref = 1;
-  string description = 2;
+  string description = 1;
   oneof transform {
     option (buf.validate.oneof).required = true;
-    TagTransform tag_transform = 3;
+    TagTransform tag_transform = 2;
   }
   message TagTransform {
+    /* 'tag_content' is the primary key of TagTransform's. It uses loose matching: dash, underscore
```
Comment (trietsch) [inline protos/getstrm/pace/api/entities/v1alpha/entities.proto]:
```suggestion
    /* 'tag_content' is the primary key of TagTransforms. It uses loose matching: dash, underscore
```

File: protos/getstrm/pace/api/entities/v1alpha/entities.proto [hunk 9]
```diff
@@ -179,29 +179,28 @@ message DataPolicy {
   }
 }
 
-// If the specified transform is incompatible with the field data type to which this global transform is connected with,
-// then the end user will get an invalid argument, and the blueprint policy with the global transforms applied won't be returned.
+/* If the specified transform is incompatible with the field data type to which this global transform is connected with,
+   then the end user will get an invalid argument, and the blueprint policy with the global transforms applied won't be returned.
+
+  Global transforms are identified by an identifier of their [transform]. These are by default loose matching, so that
```
Comment (trietsch) [inline protos/getstrm/pace/api/entities/v1alpha/entities.proto:185]: I wouldn't state anything regarding loose matching here, you've already covered that in the `TagTransform` message. We also don't know whether this also applies to other (to be added) transforms

File: examples/global-tag-transforms/global-tag-transform.yaml [hunk 10]
```diff
@@ -1,4 +1,3 @@
-ref: "pii-email"
```
Comment (trietsch) [inline examples/global-tag-transforms/global-tag-transform.yaml:1]: The `-` shouldn't be deleted.

File: protos/getstrm/pace/api/entities/v1alpha/entities.proto [hunk 11]
```diff
@@ -179,29 +179,28 @@ message DataPolicy {
...
+  'Email-PII' is matched when 'email pii' is used.
+
+ */
 message GlobalTransform {
-  // unique identifier for a certain Global Transform. When empty
-  // the identifier of the transform will be used (for TagTransform tag_content)
-  string ref = 1;
```
Comment (trietsch) [inline protos/getstrm/pace/api/entities/v1alpha/entities.proto:187]: Please also check the docs, to ensure that everything mentioned there is compatible with the removal of `ref`.

File: app/src/main/kotlin/com/getstrm/pace/util/Util.kt [hunk 12]
```diff
@@ -185,3 +179,41 @@ fun DataPolicy.Field.normalizeType(): DataPolicy.Field =
...
+                            )
+                            .build()
+                    )
+                    .build()
+            )
+        }
+    }
+}
+
+fun GlobalTransform.refAndType(): Pair<String, TransformCase> = Pair(when(transformCase){
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/util/Util.kt:207]: Good one!

File: app/src/main/kotlin/com/getstrm/pace/util/Util.kt [hunk 13]
```diff
@@ -185,3 +179,41 @@ fun DataPolicy.Field.normalizeType(): DataPolicy.Field =
...
+    TransformCase.TAG_TRANSFORM -> tagTransform.tagContent
+    TransformCase.TRANSFORM_NOT_SET -> throw BadRequestException(
+                code = BadRequestException.Code.INVALID_ARGUMENT,
+                badRequest = BadRequest.newBuilder()
+                    .addFieldViolations(
+                        BadRequest.FieldViolation.newBuilder()
+                            .setDescription( "TRANSFORM_NOT_SET")
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/util/Util.kt]: Please make this a bit more user friendly / human readable.

File: app/src/main/kotlin/com/getstrm/pace/util/Util.kt [hunk 14]
```diff
@@ -185,3 +179,41 @@ fun DataPolicy.Field.normalizeType(): DataPolicy.Field =
     toBuilder().setType(sqlDataType().typeName).build()
 
 val sqlParser = DSL.using(SQLDialect.DEFAULT).parser()
+
+/** force case type string to TransformCase
+ * don't allow TRANSFORM_NOT_SET
+ */
+fun String.toTransformCase(): TransformCase = let {
+    TransformCase.valueOf(this).also {
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/util/Util.kt:187]: The `.also` is not necessary. This is handled by protovalidate. Even the check for `TRANSFORM_NOT_SET` is not necessary, as we verify that the input type is in a specific collection. See the proto files.

File: app/src/main/kotlin/com/getstrm/pace/service/GlobalTransformsService.kt [hunk 15]
```diff
@@ -57,12 +57,8 @@ class GlobalTransformsService(
             .mapNotNull { field ->
                 val transforms = field.tagsList.flatMap { tag ->
                     // TODO this should be a batch call for multiple refAndTypes
-                    getTransformOrNull(
-                        RefAndType.newBuilder()
-                            .setRef(tag)
-                            .setType(GlobalTransform.TransformCase.TAG_TRANSFORM.name)
-                            .build()
-                    )?.tagTransform?.transformsList ?: emptyList()
+                    getTransformOrNull( tag, TransformCase.TAG_TRANSFORM )
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/service/GlobalTransformsService.kt:60]: I don't want to be nitpicking, but I do see redundant spaces. This can lead to unnecessary whitespace changes on other branches if team members touch these files. We should think of how to use ktlint / editorconfig / a different formatter properly

File: app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt [hunk 16]
```diff
@@ -16,50 +17,73 @@ import java.time.Instant
...
-
-        val record = getTransform(globalTransform.refAndType()) ?: jooq.newRecord(GLOBAL_TRANSFORMS)
-            .apply { this.createdAt = now }
+        val (ref, type) = globalTransform.refAndType()
+        val record = getTransform(ref, type) ?:
+            jooq.newRecord(GLOBAL_TRANSFORMS).apply { this.createdAt = now }
 
         record.apply {
-            this.ref = globalTransform.ref
+            /* ¡Muy importante!
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/dao/GlobalTransformsDao.kt:59]: 🇪🇸

**Review by trietsch — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/getstrm/pace/util/Util.kt (hunk 14 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/util/Util.kt:187]: But GlobalTransform is also constructed in the source code, not always via the api.
